codegen/decl.go

```diff
@@ -42,51 +42,49 @@ func (cg *CodeGen) EmitFuncDecl(ctx context.Context, scope *parser.Scope, fun *p
 
 	switch fun.Type.Primary() {
 	case parser.Filesystem:
-		return cg.EmitFilesystemBlock(ctx, fun.Scope, fun.Body.NonEmptyStmts(), ac)
+		return cg.EmitFilesystemBlock(ctx, fun.Scope, fun.Body.NonEmptyStmts())
 	case parser.Option:
-		return cg.EmitOptions(ctx, fun.Scope, string(fun.Type.Secondary()), fun.Body.NonEmptyStmts(), ac)
+		return cg.EmitOptions(ctx, fun.Scope, string(fun.Type.Secondary()), fun.Body.NonEmptyStmts())
 	case parser.Str:
 		return cg.EmitStringBlock(ctx, fun.Scope, fun.Body.NonEmptyStmts())
 	default:
 		return nil, checker.ErrInvalidTarget{Ident: fun.Name}
 	}
 }
 
-func (cg *CodeGen) EmitFilesystemFuncDecl(ctx context.Context, scope *parser.Scope, fun *parser.FuncDecl, call *parser.CallStmt, ac aliasCallback) (llb.State, error) {
-	v, err := cg.EmitFuncDecl(ctx, scope, fun, call, "", ac)
+func (cg *CodeGen) EmitFilesystemFuncDecl(ctx context.Context, scope *parser.Scope, fun *parser.FuncDecl, call *parser.CallStmt) (llb.State, error) {
+	v, err := cg.EmitFuncDecl(ctx, scope, fun, call, "")
 	if err != nil {
 		return llb.Scratch(), err
 	}
 	return v.(llb.State), nil
 }
 
 func (cg *CodeGen) EmitOptionFuncDecl(ctx context.Context, scope *parser.Scope, fun *parser.FuncDecl, call *parser.CallStmt, op string) ([]interface{}, error) {
-	v, err := cg.EmitFuncDecl(ctx, scope, fun, call, op, noopAliasCallback)
+	v, err := cg.EmitFuncDecl(ctx, scope, fun, call, op)
 	if err != nil {
 		return nil, err
 	}
 	return v.([]interface{}), nil
 }
 
-func (cg *CodeGen) EmitStringFuncDecl(ctx context.Context, scope *parser.Scope, fun *parser.FuncDecl, call *parser.CallStmt, ac aliasCallback) (string, error) {
-	v, err := cg.EmitFuncDecl(ctx, scope, fun, call, "", ac)
+func (cg *CodeGen) EmitStringFuncDecl(ctx context.Context, scope *parser.Scope, fun *parser.FuncDecl, call *parser.CallStmt) (string, error) {
+	v, err := cg.EmitFuncDecl(ctx, scope, fun, call, "")
 	if err != nil {
 		return "", err
 	}
 	return v.(string), nil
 }
 
 func (cg *CodeGen) EmitAliasDecl(ctx context.Context, scope *parser.Scope, alias *parser.AliasDecl, call *parser.CallStmt) (interface{}, error) {
-	var v interface{}
-	_, err := cg.EmitFuncDecl(ctx, scope, alias.Func, call, "", func(aliasCall *parser.CallStmt, aliasValue interface{}) {
-		if alias.Call == aliasCall {
-			v = aliasValue
-		}
-	})
+	if v, ok := cg.aliasCache.Load(alias.Call); ok {
```

Comment:
I'm not sure this is correct? What about aliases that have arguments? The cache key also needs to consider the `call.Args`.

Consider:
```hlb
fs foo(string command) {
    image "alpine"
    run string {
        format "%s > /out/foo" command
    } with option {
        dir "/src"
        mount fs { scratch; } "/out" as fooOut
}

fs default() {
    scratch
    copy fs { fooOut "echo hello"; } "/foo" "/foo1"
    copy fs { fooOut "echo world"; } "/foo" "/foo2"
}
```

The second time `fooOut` is called, the `llb.State` is cached in `aliasCache`, so both `/foo1` and `/foo2` have the content `hello`.

Reply:
Ughh, yeah, I will rethink this.